GoogleForCreators/web-stories-wp
# Checklist: Add new AMP validation check

## Context

Based on #6184 to show when there are AMP violations in a story that has been published. 

## Summary

Adds a new check for if there are AMP errors present in a story. There must be either a story link present or the status is published/scheduled. 

<img width="280" alt="Screen Shot 2021-07-28 at 2 48 36 PM" src="https://user-images.githubusercontent.com/10720454/127400686-49cdaba7-fb3a-4d39-b64f-57a572f30807.png">


## Relevant Technical Choices

- Pascal did most of this in his proof of concept (#6184), I implemented in the new checklist structure. 
- Amp Error Validation is a new **priority** check. It updates on initial render and then whenever the `isSaving` prop from the `story` becomes `false` again. Right now, the actual check for errors will only run with a story has been published or scheduled. This is an intentional choice because 'saving' fixes most if not all errors with the AMP code. 
- We are filtering out errors in the validator that are checked internally. These are: missing story poster, missing publisher logo, missing video posters. The other priority check that is covered here is the story title, but AMP allows an empty title, so our check is better and that error never sees the light of day for the validator (as far as I can tell) 
- I have put this check behind its own feature flag because I'm not sure all of the issues we're filtering for are covered yet and it feels like something we might want to have the ability to exclude. Experiment is called `enableAmpValidationCheck`, set it to default true.


Fix three small bugs in checklist while making sense of the logic here: 
- the `storyPageCount` was digging one level too deep for `pages` and never rendering - pretty sure this was just a sneaky rebase
- publish/update button was keeping the exclamation icon and tooltip even when all checkpoints were hit and issues resolved. Limiting based on if the dialog should appear when checkpoint ALL is hit (also updating the `null` to an empty string for tooltip to silence a console warning about the propType. 
- update Panel sections to only set `isExpanded` to true when there's an accompanying issue count and the `isOpen` boolean is true, this was causing a little hiccup where if a section was expanded when all issues were solved and now the checklist is empty the helper text associated with that panel wouldn't be hidden. 
<img width="379" alt="Screen Shot 2021-07-28 at 11 23 53 AM" src="https://user-images.githubusercontent.com/10720454/127396470-cce39521-5fc3-403d-9bc2-04b7857f5c93.png">



## To-do

- [x] Need to add some tests for the check
- Look into Web Assembly implementation (new ticket here: https://github.com/google/web-stories-wp/issues/8477)
- [x] Finalize copy
- [ ] poster images for videos should trigger an error but can't find it! 

## User-facing changes

If you have AMP errors in your published/scheduled story you see this: 

<img width="286" alt="Screen Shot 2021-07-29 at 12 06 23 PM" src="https://user-images.githubusercontent.com/10720454/127551386-913be363-3c97-4aee-9d07-88239cc99845.png">




## Testing Instructions

Have a bad story and see the error. This is surprisingly hard to trigger. 
this is a list of AMP errors: https://amp.dev/documentation/guides-and-tutorials/learn/validation-workflow/validation_errors/?format=websites#amp-html-tag-and-attribute-errors  if you find a way to get this to fire please share! I was having luck before moving the check to be post publish. All errors I've found get resolved on save. 

### QA






This PR can be tested by following these steps:

1.

### UAT







This PR can be tested by following these steps:

1.

## Reviews

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

- I added a new `trackClick` (`click_checklist_amp_test`) for the link that sends users to `https://search.google.com/test/amp` to test their story and get more robust details about what went wrong in their story. I figure this is useful to see if people are following through on testing stories with issues or if we need a more involved solution for walking users through fixing these errors. 
- Also added a new `trackEvent` (`amp_validation_errors_found`) which fires once per session when amp validation errors are found. 

### Does this PR have a legal-related impact?

No

## Checklist




---



Fixes #8442
